### PR TITLE
gonum: add Version query function

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Updates #260.

### CODE REVIEW

**Comment by kortschak** [comment]
Are we OK with the format of the versions returned? Do we want to warn that this may change?

**Comment by kortschak** [comment]
*ping*

File: version.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	for _, m := range b.Deps {
+		if m.Path == root {
+			if m.Replace != nil {
+				switch {
+				case m.Replace.Version != "" && m.Replace.Path != "":
+					return fmt.Sprintf("%s %s", m.Replace.Path, m.Replace.Version), m.Replace.Sum
+				case m.Replace.Version != "":
+					return m.Replace.Version, m.Replace.Sum
+				case m.Replace.Path != "":
+					return m.Replace.Path, m.Replace.Sum
```
Comment (sbinet) [inline version.go]: shouldn't this be `return fmt.Sprintf("%s %s*", m.Replace.Path, m.Version), m.Replace.Sum` instead?

File: version.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+				default:
+					return m.Version + "*", ""
```
Comment (sbinet) [inline version.go]: `s/return m.Version + "*", ""/return m.Version + "*", m.Sum+"*"/` perhaps?

File: version.go (hunk 1 above)
Comment (kortschak) [reply version.go]: I think so, if `m.Version != ""`.

File: version.go (hunk 2 above)
Comment (kortschak) [reply version.go]: OK

**Comment by kortschak** [comment]
PTAL

**Comment by kortschak** [comment]
I have a little module in my local called `latest_gonum`. This could be place either as a module in gonum/gonum or could be put in gonum/tools. What do you think?

**Comment by sbinet** [comment]
how would that work? it seems a bit cumbersome to anticipate the checksum of the whole `gonum/gonum` repo...
(hence my suggestion of breaking `Version()` into `Version() + versionOf(b)` so we could write tests using `versionOf(b)` with a manufactured `*debug.BuildInfo` we control...)

**Comment by kortschak** [comment]
It's not for testing. It's a command that can be used to find the latest version. The version doesn't need to be anticipated, the go.mod can be artificially created with `latest` as the version.

I didn't break for testing because the complexity is not in the switch, but rather then `debug.BuildInfo` structure that `debug.ReadBuildInfo` can return. That's outside of our knowledge.

**Comment by sbinet** [comment]
> It's not for testing. It's a command that can be used to find the latest version.

ah. ok. on the basis that it would be a bit silly to install `gonum.org/v1/gonum/cmd/latest_gonum` to get that binary (but then, one could just look into the repo one had just installed), I'd err on putting it under `gonum/tools`.
but that's really just to break the tie.

> I didn't break for testing because the complexity is not in the switch, but rather then debug.BuildInfo structure that debug.ReadBuildInfo can return. That's outside of our knowledge.

yes, but at least this would help understand what's going one if/when `ReadBuildInfo` returns some slightly different output. (under `Go1` compatibility rules that shouldn't happen, of course...)
but ok, this can be left as is.

**Comment by kortschak** [comment]
> on the basis that it would be a bit silly to install gonum.org/v1/gonum/cmd/latest_gonum to get that binary

These would have the same effect. The `latest_gonum` module itself would be small, but would pull in the gonum/gonum module. But this is true also of the gonum/tools module approach.

**Pull request merged**